kernel/src/table_properties.rs

```diff
@@ -0,0 +1,299 @@
+            Error::generic("couldn't parse DataSkippingNumIndexedCols to an integer")
+        })?;
+        match num {
+            -1 => Ok(DataSkippingNumIndexedCols::AllColumns),
+            x => Ok(DataSkippingNumIndexedCols::NumColumns(
+                x.try_into().map_err(|_| {
+                    Error::generic("couldn't parse DataSkippingNumIndexedCols to positive integer")
+                })?,
+            )),
+        }
+    }
+}
+
+/// The isolation level applied during transaction
+#[derive(Debug, EnumString, Default, Copy, Clone, PartialEq, Eq)]
+#[strum(serialize_all = "camelCase")]
+pub enum IsolationLevel {
+    /// The strongest isolation level. It ensures that committed write operations
+    /// and all reads are Serializable. Operations are allowed as long as there
+    /// exists a serial sequence of executing them one-at-a-time that generates
+    /// the same outcome as that seen in the table. For the write operations,
+    /// the serial sequence is exactly the same as that seen in the table’s history.
+    #[default]
+    Serializable,
+
+    /// A weaker isolation level than Serializable. It ensures only that the write
+    /// operations (that is, not reads) are serializable. However, this is still stronger
+    /// than Snapshot isolation. WriteSerializable is the default isolation level because
+    /// it provides great balance of data consistency and availability for most common operations.
+    WriteSerializable,
+
+    /// SnapshotIsolation is a guarantee that all reads made in a transaction will see a consistent
+    /// snapshot of the database (in practice it reads the last committed values that existed at the
+    /// time it started), and the transaction itself will successfully commit only if no updates
+    /// it has made conflict with any concurrent updates made since that snapshot.
+    SnapshotIsolation,
+}
+
+/// The checkpoint policy applied when writing checkpoints
+#[derive(Debug, EnumString, Default, Clone, PartialEq, Eq)]
+#[strum(serialize_all = "camelCase")]
+pub enum CheckpointPolicy {
+    /// classic Delta Lake checkpoints
+    #[default]
+    Classic,
+    /// v2 checkpoints
+    V2,
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use crate::expressions::column_name;
+    use std::collections::HashMap;
+
+    #[test]
+    fn known_key_unknown_val() {
+        let properties = HashMap::from([("delta.appendOnly".to_string(), "wack".to_string())]);
+        let table_properties = TableProperties::from(properties.iter());
+        let unknown_properties =
+            HashMap::from([("delta.appendOnly".to_string(), "wack".to_string())]);
+        let expected = TableProperties {
+            unknown_properties,
+            ..Default::default()
+        };
+        assert_eq!(table_properties, expected);
+    }
+
+    #[test]
+    fn allow_unknown_keys() {
+        let properties = [("some_random_unknown_key".to_string(), "test".to_string())];
```

Comment:
Could be "fun" to do:
```suggestion
        let properties = [("unknown_properties".to_string(), "messing with your mind".to_string())];
```

Reply:
haha will add!

Reply:
did something slightly shorter to keep on one line